libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1718 @@
+        } else {
+            self.tokens.push(Token{ token_type: token, text: None })
+        }
+        self.current_string = String::new();
+    }
+
+    // ignore empties self.current_string. This is used for whitespace tokens since they are not currently
+    // taken into consideration when parsing
+    fn ignore(&mut self) {
+        self.current_string = String::new();
+    }
+
+    fn keyword_or_ident(&mut self) -> TokenType {
+        while let Some(letter) = self.source.peek() {
+            if letter.is_alphanumeric() {
+                self.next();
+                continue; 
+            }
+            break; // if letter is not alphanumeric
+        }
+
+        let current: &str = &self.current_string; 
+        match current {
+            "forall" => TokenType::FORALL, 
+            "where" => TokenType::WHERE, 
+            "int" => TokenType::INT, 
+            "float" => TokenType::FLOAT, 
+            "string" => TokenType::STRING, 
+            "bool" => TokenType::BOOL, 
+            "uint" => TokenType::UINT, 
+            "duration" => TokenType::DURATION, 
+            "time" => TokenType::TIME, 
+            "regexp" => TokenType::REGEXP, 
+            _ => TokenType::IDENTIFIER
+        }
+    }
+}
+
+// lex_token recursively lexes characters using StateFn
+fn lex_token(lexer: &mut Lexer) -> Option<TokenType> {
+    match lexer.next() {
+        Some(letter) if letter.is_alphanumeric() =>  Some(lexer.keyword_or_ident()), 
+        Some(letter) if letter.is_whitespace() => Some(TokenType::WHITESPACE), 
+        Some(letter) if letter == '{' => Some(TokenType::LEFTCURLYBRAC), 
+        Some(letter) if letter == '}' => Some(TokenType::RIGHTCURLYBRAC), 
+        Some(letter) if letter == '[' => Some(TokenType::LEFTSQUAREBRAC), 
+        Some(letter) if letter == ']' => Some(TokenType::RIGHTSQUAREBRAC), 
+        Some(letter) if letter == '(' => Some(TokenType::LEFTPAREN), 
+        Some(letter) if letter == ')' => Some(TokenType::RIGHTPAREN), 
+        Some(letter) if letter == '?' => Some(TokenType::QUESTIONMARK), 
+        Some(letter) if letter == ':' => Some(TokenType::COLON), 
+        Some(letter) if letter == ',' => Some(TokenType::COMMA), 
+        Some(letter) if letter == '+' => Some(TokenType::PLUS), 
+        Some(letter) if letter == '<' => {
+            match lexer.next() {
+                Some('-') => Some(TokenType::PIPE),
+                _ => Some(TokenType::ERROR), 
+            }
+        },
+        Some(letter) if letter == '-' => {
+            match lexer.next() {
+                Some('>') => Some(TokenType::ARROW), 
+                _ => Some(TokenType::ERROR), 
+            }
+        },
+        Some(_) => Some(TokenType::ERROR), 
+        _ => Some(TokenType::EOF), 
+    }
+}
+
+/** PARSER */
+
+struct Parser<'a> {
+    tokens: Peekable<Iter<'a, Token>>,
+}
+
+// parse_tokens is a testing function for directly parsing a vector of tokens. 
```

Comment:
I thought it might have been useful to have a few tests that just focused on the Lexer or the Parser. Testing parse tests them both together. I'll also need to refactor several of my tests.

Reply:
I can go ahead and change this. We just lose the ability to test the parser in isolation.